LiS/man/man1/oc.1.man
'\" rtp
.\" -*- nroff -*- vim: ft=nroff nocin nosi
.\"
.\" @(#) $Id$
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2008-2009  Monavacon Limited <http://www.monavacon.com/>
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\" Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to make and distribute verbatim copies of this
.\" manual provided the copyright notice and this permission notice are
.\" preserved on all copies.
.\"
.\" Permission is granted to copy and distribute modified versions of this
.\" manual under the conditions for verbatim copying, provided that the
.\" entire resulting derived work is distributed under the terms of a
.\" permission notice identical to this one
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this
.\" manual page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from
.\" the use of the information contained herein.  The author(s) may not
.\" have taken the same level of care in the production of this manual,
.\" which is licensed free of charge, as they might when working
.\" professionally.  The author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by
.\" the source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software
.\" on behalf of the U.S. Government ("Government"), the following
.\" provisions apply to you.  If the Software is supplied by the Department
.\" of Defense ("DoD"), it is classified as "Commercial Computer Software"
.\" under paragraph 252.227-7014 of the DoD Supplement to the Federal
.\" Acquisition Regulations ("DFARS") (or any successor regulations) and the
.\" Government is acquiring only the license rights granted herein (the
.\" license rights customarily provided to non-Government users).  If the
.\" Software is supplied to any unit or agency of the Government other than
.\" DoD, it is classified as "Restricted Computer Software" and the
.\" Government's rights in the Software are defined in paragraph 52.227-19
.\" of the Federal Acquisition Regulations ("FAR") (or any successor
.\" regulations) or, in the cases of NASA, in paragraph 18.52.227-86 of the
.\" NASA Supplement to the FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.\" Last Modified $Date$ by $Author$
.\"
.\" -------------------------------------------------------------------------
.\"
.\" $Log: oc.1.man,v $
.\" Revision 1.1.4.4  2008-04-29 08:33:17  brian
.\" - update headers for Affero release
.\"
.\" Revision 1.1.4.3  2006/04/05 09:42:17  brian
.\" - moved .so line and updated headers
.\"
.\" =========================================================================
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database lis.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
.R2
.so lis.macros
.\"
.\"
.TH OC 1 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ User Commands"
.\"
.\"
.SH NAME
.B oc
\- open and close testing of STREAMS files.
.\"
.\"
.SH SYNOPSIS
.PP
.HP 12
\fBoc\fR [\fIoptions\fR] \fIPATH\fP ...
.PD 0
.HP 12
\fBoc\fR {\fB-h\fR|\fB--help\fR}
.HP 12
\fBoc\fR {\fB-V\fR|\fB--version\fR}
.HP 12
\fBoc\fR {\fB-C\fR|\fB--copying\fR}
.PD
.\"
.\"
.SH ARGUMENTS
.PP
.TP
.I PATH
The path or paths to STREAMS devices to open (and close).
.\"
.\"
.SH OPTIONS
.PP
.TP
\fB-n\fP, \fB--number\fP \fINUMBER\fP
set the number of iterations.
.TP
\fB-w\fP, \fB--wait\fP \fIWAIT\fP
set the close wait in milliseconds.  The default value is zero (\fB0\fP).
.TP
\fB-r\fP, \fB--recvfd\fP \fIRECVFD\fP
set the receive file descriptor.
.TP
\fB-W\fP, \fB--wronly\fP
open write only.
.TP
\fB-R\fP, \fB--rdonly\fP
open read only.
.TP
\fB-B\fP, \fB--nonblock\fP
open for non-blocking operation.
.TP
\fB-v\fP, \fB--verbose\fP [\fILEVEL\fP]
increase verbosity or set to \fILEVEL\fP.  The default value is one (\fB1\fP).
.TP
\fB-q\fP, \fB--quiet\fP
suppress normal output (equivalent to \fB--verbose\fP=\fI0\fP).
.TP
\fB-h\fP, \fB--help\fP, \fB-?\fP, \fB--?\fP
print this usage message and exit.
.TP
\fB-V\fP, \fB--version\fP
print version and exit.
.TP
\fB-C\fP, \fB--copying\fP
print copying permission and exit.
.\"
.\"
.SH DESCRIPTION
.PP
Given a list of paths in the \fIPATH\fP arguments, this problem opens the paths a number of times
each as specified by the \fB-n\fP option, and then waits an optional number of seconds as specified
by the \fB-w\fP option, and then exits.
.\"
.\"
.\".SH DIAGNOSTICS
.\".PP
.\"
.\"
.\".SH ERRORS
.\".PP
.\"
.\"
.SH NOTICES
.PP
.B oc
is normally invoked from the
.B @PACKAGE@-@VERSION@
test suite.
.\"
.\"
.SH FILES
.PP
.TP
.B /usr/libexec/LiS/oc
contains the
.B oc
command.
.\"
.\"
.SH "SEE ALSO"
.PP
.BR rwf (1),
.BR pafd (1),
.BR psfd (1),
.BR test-streams (8),
.BR STREAMS (4),
.IR "Linux STREAMS Installation and Reference Manual" .
.\"
.\"
.SH BUGS
.PP
.B oc
has no known bugs.
.\"
.\"
.\".SH COMPATIBILITY
.\".PP
.\"
.\"
.SH CONFORMANCE
.PP
.B oc
tests conformance of
.BR STREAMS (4)
to the specifications listed on the
.BR STREAMS (4)
manual page.
.\"
.\"
.SH HISTORY
.PP
.B oc
is provided with the
.I @PACKAGE_TITLE@
package.
.\"
.\"
.[
$LIST$
.]
.TI
.\" vim: ft=nroff sw=4 noet nocin nosi com=b\:.\\\" fo+=tcqlorn tw=77


NEWS
OpenSS7 Master Package -- history of user visible changes.  2009-05-04
$Id: NEWS,v 0.9.2.72 2009-04-19 00:28:25 brian Exp $
Copyright (c) 2008-2009  Monavacon Limited. <http://www.monavacon.com/>
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2001  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
`OpenSS7 Master Package' package.

 Major changes for release openss7-0.9.2.F.20090405
---------------------------------------------------

This is the seventh `OpenSS7 Project' release of the `OpenSS7 Master
Package' package.  For a sub-package listing, see `README'
(http://www.openss7.org/codefiles/openss7/README) in the release.  Most
of the sub-packages in the release are production grade for `Linux
Fast-STREAMS'.(1)  All existing validation test suites run clean on
supported distributions, production kernels and architectures.

The original intention was that the `OpenSS7 Master Package' package
not be released as often as subpackages and, indeed, newer subpackages
can be strapped into the master package and built automatically.
However, in reality, the `OpenSS7 Master Package' package has been
released as often as the subpackages.  To rebuild the master package
with a new sub-package release, simply replace the directory to which
the sub-package belongs with the unpackged sub-package reelase and then
rebuild the master package.  This release provides support for recent
distributions, production kernels and tool chains.  The release also
provides support for significiant new features in the sub-packages.

Major features since the last public release are:

   - Deprecated sub-packages have been excluded from the master package
     release.  The excluded sub-packages are:

         - `sctp-0.2.27.20090405'
         - `iperf-2.0.8.20090405'
         - `netperf-2.3.7.20090405'
         - `osr61-0.9.2.3.20090405'

   - Documentation license upgraded to GFDL 1.3.

   - Updated tool chain to `m4-1.4.13', `automake-1.10.2',
     `bison-2.4.1', `libtool-2.2.6a', `texinfo-4.13a'.

   - Addition of the following sub-packages:

         - `stratm-0.9.2.1.20090405'

   - Addition of the `stratm' sub-package to the master build and
     release that now includes all of the ATM components that were
     previously part of the `strss7' sub-package.

   - Significant feature updates were made to:

         - `strchan-0.9.2.4.20090405'
         - `strsock-0.9.2.4.20090405'
         - `strnsl-0.9.2.4.20090405'
         - `sigtran-0.9.2.4.20090405'

   - Significant rework of the `strchan' sub-package into the `strss7',
     `strisdn', `strx25' and `striso' packages.  One single driver now
     meets the needs of all sub-packages.

   - Completion of the `strsock' package, moves from pre-alpha to beta.

   - The `strnsl' subpackage moves from alpha to production.

   - Significant completion of the `sigtran' package: all components
     moved back to production.


_This is a public, stable, production grade release of the package: it
deprecates previous releases.  Please upgrade to the current release
before reporting bugs._

As with other OpenSS7 releases, this release configures, compiles,
installs and builds RPMs and DEBs for a wide range of Linux 2.4 and 2.6
`rpm(8)' and `dpkg(1)' based distributions, and can be used on
production kernels without patching or recompiling the kernel.

This package is publicly released strictly under the `GNU Affero
General Public License Version 3'; however, commercial licensing is
available.  The release is available as an `autoconf(1)' tarball, SRPM,
DSC, and a set of binary RPMs and DEBs for popular distributions.  See
the package page (http://www.openss.org/openss7_pkg.html) for the
`autoconf(1)' tarballs, SRPMs, DSCs and binary RPMs and DEBs.  See the
repository page (http://www.openss7.org/repos.html) for information
concerning network installation and update sources.

See
`http://www.openss7.org/codefiles/openss7-0.9.2.F.20090405/ChangeLog'
and `http://www.openss7.org/codefiles/openss7-0.9.2.F.20090405/NEWS' in
the release for more information.  Also, see the `openss7.pdf' manual
in the release (also in html
(http://www.openss7.org/openss7_manual.html)).

For the news release, see `http://www.openss7.org/rel2009XXXX_L.html'.

---------- Footnotes ----------

(1) Sub-packages are no longer supported on the deprecated `LiS' STREAMS
package.

 Major changes for release openss7-0.9.2.G
------------------------------------------

This is the sixth public release of the OpenSS7 Master Package.  See
`README' in the release for a sub-package listing.  Most of the
sub-packages in the release are production grade for `Linux
Fast-STREAMS'.  All existing validation test suites run clean on
supported distributions and architectures.

The `OpenSS7 Master Package' is not released as often as the
sub-packages.  As sub-packages are released more often, to rebuild the
master package with a new sub-package release, simply replace the
directory to which the sub-package belongs with the unpacked
sub-package release and then rebuild the master package.  This release
provides support for recent distributions and tool chains.

Major features since the last public release are as follows:

   - License upgrade to `AGPL Version 3'.

   - Modifications to build under `Fedora' `2.6.22.5-49' kernel.  These
     changes also support `2.6.22.9-91.fc7' kernel.  Modifications to
     build under `Fedora' `2.6.25-45.fc9' and `2.6.26.5-45.fc9' kernels.
     Documented `lib32gcc1' problem on Ubuntu.  Noted problem running
     under XEN kernels.  XEN kernels are not yet supported.  Added
     `MODULE_VERSION' to all modules and drivers.

   - Ability to strap out major documentation build and installation
     primarily for embedded targets.  Improvements to common build
     process for embedded and cross-compile targets.  Cross-compile

LiS/util/linux/oc.c
/*****************************************************************************

 @(#) $RCSfile: oc.c,v $ $Name:  $($Revision: 1.1.1.1.12.9 $) $Date: 2008-04-29 08:33:28 $

 -----------------------------------------------------------------------------

 Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2000  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software: you can redistribute it and/or modify it under
 the terms of the GNU General Public License as published by the Free Software
 Foundation, version 3 of the license.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 details.

 You should have received a copy of the GNU General Public License along with
 this program.  If not, see <http://www.gnu.org/licenses/>, or write to the
 Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 -----------------------------------------------------------------------------

 Last Modified $Date: 2008-04-29 08:33:28 $ by $Author: brian $

 -----------------------------------------------------------------------------

 $Log: oc.c,v $
 Revision 1.1.1.1.12.9  2008-04-29 08:33:28  brian
 - update headers for Affero release

 Revision 1.1.1.1.12.8  2007/08/14 10:47:33  brian
 - GPLv3 header update

 Revision 1.1.1.1.12.5  2005/07/13 12:01:24  brian
 - working up compat and check pass (finally lindented LiS)

 Revision 1.1.1.1.12.3  2005/05/14 08:35:16  brian
 - copyright header correction

 *****************************************************************************/

#ident "@(#) $RCSfile: oc.c,v $ $Name:  $($Revision: 1.1.1.1.12.9 $) $Date: 2008-04-29 08:33:28 $"

static char const ident[] =
    "$RCSfile: oc.c,v $ $Name:  $($Revision: 1.1.1.1.12.9 $) $Date: 2008-04-29 08:33:28 $";

/* 
 *  oc.c - open/<default>close testing of files.  Given a list of paths,
 *         this program simply opens them an optional number of times each,
 *         then waits an optional number of seconds before exiting.
 * 
 *         This program also supports receiving passed file descriptors.
 *
 *  Copyright (C) 2000  John A. Boyd Jr.  protologos, LLC
 */

#include <unistd.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <memory.h>
#include <fcntl.h>
#include <sys/poll.h>
#ifdef _GNU_SOURCE
#include <getopt.h>
#endif				/* _GNU_SOURCE */
#include <stropts.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/sysmacros.h>
#include <sys/ioctl.h>
#include <errno.h>

int verbose = 1;

void
copying(int argc, char *argv[])
{
	if (!verbose)
		return;
	fprintf(stdout, "\
\n\
%1$s %2$s:\n\
\n\
Copyright (c) 2003-2008  OpenSS7 Corporation <http://www.openss7.com/>\n\
Copyright (c) 2000       John A. Boyd Jr.  protologos, LLC\n\
\n\
All Rights Reserved.\n\
\n\
This program is free software;  you can  redistribute  it and/or modify it under\n\
the terms of the GNU General  Public License as  published by the  Free Software\n\
Foundation; version 3 of the License.\n\
\n\
This program is distributed in the hope that it will be  useful, but WITHOUT ANY\n\
WARRANTY;  without even the implied warranty of MERCHANTABILITY or FITNESS FOR A\n\
PARTICULAR PURPOSE.  See the GNU General Public License for more details.\n\
\n\
You should  have received  a copy of the GNU  General  Public License along with\n\
this program.  If not, see <http://www.gnu.org/licenses/>,  or write to the Free\n\
Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.\n\
\n\
U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on behalf\n\
of the  U.S. Government  (\"Government\"),  the following provisions apply to you.\n\
If the Software is  supplied by the Department of Defense (\"DoD\"), it is classi-\n\
fied as  \"Commercial Computer Software\"  under paragraph 252.227-7014 of the DoD\n\
Supplement  to the  Federal Acquisition Regulations  (\"DFARS\") (or any successor\n\
regulations) and the  Government  is acquiring  only the license rights  granted\n\
herein (the license  rights customarily  provided to non-Government  users).  If\n\
the Software is supplied to any unit or agency of the Government other than DoD,\n\
it is classified as  \"Restricted Computer Software\" and the  Government's rights\n\
in the  Software are defined in  paragraph 52.227-19 of the Federal  Acquisition\n\
Regulations  (\"FAR\") (or any successor regulations) or, in the cases of NASA, in\n\
paragraph  18.52.227-86 of the  NASA Supplement  to the  FAR (or  any  successor\n\
regulations).\n\
\n\
", argv[0], ident);
}

void
version(int argc, char *argv[])
{
	if (!verbose)
		return;
	fprintf(stdout, "\
\n\
%1$s %2$s:\n\
    Copyright (c) 2003-2008  OpenSS7 Corporation.  All Rights Reserved.\n\
    Copyright (c) 2000       John A. Boyd Jr.  protologos, LLC\n\
\n\
    Distributed by OpenSS7 Corporation under GPL Version 3,\n\
    included here by reference.\n\
\n\
    See `%1$s --copying' for copying permissions.\n\
\n\
", argv[0], ident);
}

void
usage(int argc, char *argv[])
{
	if (!verbose)
		return;
	fprintf(stderr, "\
Usage:\n\
    %1$s [options] PATH ...\n\
    %1$s {-h,--help}\n\
    %1$s {-V,--version}\n\
    %1$s {-C,--copying}\n\
", argv[0]);
}

void
help(int argc, char *argv[])
{
	if (!verbose)
		return;
	fprintf(stdout, "\
\n\
Usage:\n\
    %1$s [options] PATH ...\n\
    %1$s {-h,--help}\n\
    %1$s {-V,--version}\n\
    %1$s {-C,--copying}\n\
Arguments:\n\
    PATH ...\n\
        The path or paths to STREAMS devices open (and close).\n\
Options:\n\
    -n, --number NUMBER\n\
        set the number of iterations\n\
    -w, --wait WAIT\n\
        set the close wait in milliseconds [default: 0]\n\
    -r, --recvfd RECVFD\n\
        set the receive file descriptor\n\
    -W, --wronly\n\
        open write only\n\
    -R, --rdonly\n\
        open read only\n\
    -B, --nonblock\n\
        open for non-blocking operation\n\
    -v, --verbose [LEVEL]\n\
        increase verbosity or set to LEVEL [default: 1]\n\
        this option may be repeated.\n\
    -q, --quiet\n\
        suppress normal output (equivalent to --verbose=0)\n\
    -h, --help, -?, --?\n\
        print this usage message and exit\n\
    -V, --version\n\
        print version and exit\n\
    -C, --copying\n\
        print copying permission and exit\n\
\n\
", argv[0]);
}

int
main(int argc, char *argv[])
{
	char path[40];
	int fd, i, n = 1;
	struct stat stat;
	int wt = 0;
	int o_flags, o_wronly = 0, o_rdonly = 0, o_nonblock = 0;
	int is_stream = 0, recv_fd = 0;
	int p0;

	for (;;) {
		int c;

#ifdef _GNU_SOURCE
		int option_index = 0;
		/* *INDENT-OFF* */
		static struct option long_options[] = {
			{ "number",	required_argument,  NULL, 'n' },
			{ "wait",	required_argument,  NULL, 'w' },
			{ "recvfd",	no_argument,	    NULL, 'r' },
			{ "wronly",	required_argument,  NULL, 'W' },
			{ "rdonly",	no_argument,	    NULL, 'R' },
			{ "nonblock",	no_argument,	    NULL, 'B' },
			{ "quiet",	no_argument,	    NULL, 'q' },
			{ "verbose",	optional_argument,  NULL, 'v' },
			{ "help",	no_argument,	    NULL, 'h' },
			{ "version",	no_argument,	    NULL, 'V' },
			{ "copying",	no_argument,	    NULL, 'C' },
			{ "?",		no_argument,	    NULL, 'h' },
			{ 0, }
		};
		/* *INDENT-ON* */

		c = getopt_long_only(argc, argv, "n:w:rWRBqv::hVC?", long_options, &option_index);
#else				/* _GNU_SOURCE */
		c = getopt(argc, argv, "n:w:rWRBqv::hVC?");
#endif				/* _GNU_SOURCE */
		if (c == -1)
			break;
		switch (c) {
		case 'n':	/* -n, --number */
			n = strtol(optarg, NULL, 0);
			break;
		case 'w':	/* -w, --wait */
			wt = strtol(optarg, NULL, 0);
			break;
		case 'r':	/* -r, --recvfd */
			recv_fd = 1;
			break;
		case 'W':	/* -W, --wronly */
			o_wronly = 1;
			o_rdonly = 0;
			break;
		case 'R':	/* -R, --rdonly */
			o_rdonly = 1;
			o_wronly = 0;
			break;
		case 'B':	/* -B, --nonblock */
			o_nonblock = 1;
			break;
		case 'q':	/* -q, --quiet */
			verbose = 0;
			break;
		case 'v':	/* -v, --verbose [LEVEL] */
			if (!optarg)
				verbose++;
			else
				verbose = strtol(optarg, NULL, 0);
			break;
		case 'h':	/* -h, --help, -?, --? */
			help(argc, argv);
			exit(0);
		case 'V':	/* -V, --version */
			version(argc, argv);
			exit(0);
		case 'C':	/* -C, --copying */
			copying(argc, argv);
			exit(0);
		case '?':
		default:
			optind--;
			if (optind < argc && verbose) {
				fprintf(stderr, "%s: illegal syntax -- ", argv[0]);
				for (; optind < argc; optind++)
					fprintf(stderr, "%s ", argv[optind]);
				fprintf(stderr, "\n");
			}
			usage(argc, argv);
			exit(2);
		}
	}
	o_flags =
	    (o_wronly ? O_WRONLY : (o_rdonly ? O_RDONLY : O_RDWR)) | (o_nonblock ? O_NONBLOCK : 0);

	if (n <= 1)
		n = 1;

	for (i = 0; i < n; i++) {
		p0 = optind;
		while (argc > p0) {
			strcpy(path, argv[p0++]);

			if ((fd = open(path, o_flags)) < 0) {
				fprintf(stderr, "open( \"%s\", 0%o ) failed: %s (%d)\n", path,
					o_flags, strerror(errno), errno);
				exit(1);
			}

			if (verbose) {
				fstat(fd, &stat);
				printf("%d: fd %d = open( \"%s\", 0%o ) "
				       "mode 0%o dev 0x%04x rdev 0x%04x", i, fd, path, o_flags,
				       (int) (stat.st_mode), (int) (stat.st_dev),
				       (int) (stat.st_rdev));
				is_stream = isastream(fd);
				printf("%s", (is_stream ? " [STREAM]" : ""));
				printf("\n");
				fflush(stdout);
			}

			if (is_stream && recv_fd) {
				struct pollfd pfd;
				struct strrecvfd recv;

				/* 
				 *  FIXME: poll wt*1000 for input, and if OK try to
				 *  ioctl( I_RECVFD ).  If OK, report the received
				 *  file via stat.
				 */
				pfd.fd = fd;
				pfd.events = POLLIN;
				if (wt && !poll(&pfd, 1, wt * 1000)) {
					/* 
					 *  poll timed out
					 */
					if (verbose)
						fprintf(stderr,
							"poll(..., %dms) timed out waiting for RECVFD\n",
							wt * 1000);
				} else {
					if (ioctl(fd, I_RECVFD, &recv) < 0) {
						if (verbose)
							fprintf(stderr,
								"ioctl(%d, I_RECVFD,...) failed: %s\n",
								fd, strerror(errno));
					} else {
						if (verbose) {
							fstat(recv.fd, &stat);
							printf("%d: fd %d = received"
							       " (mode 0%o dev 0x%04x rdev 0x%04x)",
							       i, recv.fd, (int) (stat.st_mode),
							       (int) (stat.st_dev),
							       (int) (stat.st_rdev));
							printf("%s",
							       (isastream(recv.fd) ? " [STREAM]" :
								""));
							printf("\n");
							fflush(stdout);
						}
					}
				}
			}
		}
	}
	if (wt > 0) {
		while (wt--) {
			if (verbose)
				fprintf(stderr, ".");
			sleep(1);
		}
		if (verbose)
			fprintf(stderr, "!\n");
	}

	exit(0);
}
